File: plugins/cbre-il/skills/cbre-property-longlist/evals/chrome_v45_test.mjs
// evals/chrome_v45_test.mjs - execute the REAL v45 chrome from a BUILT dashboard.
//
// The five behaviours here are the ones a structural grep cannot tell apart from a comment
// that mentions them, so they are EXECUTED:
//   1. titleStr()   - the client's own displayName wins over the park + unit composition.
//   2. partyLine()  - the card names the party, landlord first, and the label follows the
//                     VALUE so a developer is never printed under a "Landlord" label.
//   3. the FOUR RENT rows always print (BLANK included) while every other row still omits.
//   4. the modal's meta chips are sentinel-GUARDED, and a link chip is emitted only for a
//                     stated http(s) URL - never for the blank token, which is a string and
//                     would otherwise have shipped href="TBC" as a live link.
//   5. Compare cells carry no highlight class.
// Offline; no npm deps.  Usage: node chrome_v45_test.mjs <built_html_path>
import fs from 'node:fs';
import vm from 'node:vm';

const htmlPath = process.argv[2];
if (!htmlPath) { console.error('usage: node chrome_v45_test.mjs <built_html>'); process.exit(2); }
const html = fs.readFileSync(htmlPath, 'utf8');

const scripts = [...html.matchAll(/<script(?![^>]*\bsrc=)[^>]*>([\s\S]*?)<\/script>/gi)].map(m => m[1]);
const code = scripts.join('\n;\n') +
  '\n;\n__capture__("PROPS", typeof PROPS !== "undefined" ? PROPS : undefined);' +
  '\n__capture__("BLANK", typeof BLANK !== "undefined" ? BLANK : undefined);' +
  '\n__capture__("titleStr", typeof titleStr !== "undefined" ? titleStr : undefined);' +
  '\n__capture__("partyLine", typeof partyLine !== "undefined" ? partyLine : undefined);' +
  '\n__capture__("cardHTML", typeof cardHTML !== "undefined" ? cardHTML : undefined);' +
  '\n__capture__("detailHTML", typeof detailHTML !== "undefined" ? detailHTML : undefined);\n';

// See modal_render_test.mjs for why this sandbox is shaped the way it is.
const sink = new Proxy(function () {}, {
  get: (_t, p) => (p === Symbol.toPrimitive || p === 'toString' || p === 'valueOf') ? () => '' : sink,
  apply: () => sink, construct: () => sink, has: () => true,
});
const target = { console };
for (const name of Object.getOwnPropertyNames(globalThis)) {
  if (!(name in target)) { try { target[name] = globalThis[name]; } catch { /* ignore */ } }
}
target.globalThis = target;
target.__capture__ = (name, val) => { target[name] = val; };
const ctx = vm.createContext(new Proxy(target, {
  get: (t, p) => (p in t ? t[p] : sink),
  has: () => true,
}));

try { vm.runInContext(code, ctx, { filename: 'built.inline.js' }); }
catch (e) { console.error('FAIL: template script threw during eval:', e && e.message); process.exit(1); }

const { PROPS: props, BLANK, titleStr, partyLine, cardHTML, detailHTML } = target;
for (const [n, v] of Object.entries({ titleStr, partyLine, cardHTML, detailHTML })) {
  if (typeof v !== 'function') { console.error(`FAIL: could not capture ${n}() from the built chrome`); process.exit(1); }
}
if (!Array.isArray(props)) { console.error('FAIL: could not capture PROPS'); process.exit(1); }

const fails = [];
const ck = (ok, label) => { console.log((ok ? '  ok   ' : '  FAIL ') + label); if (!ok) fails.push(label); };
const byPark = p => props.find(x => x.park === p);

console.log('-- BLANK: one token, and it is the one the pipeline writes --');
ck(BLANK === 'TBC', `the chrome's BLANK is ${JSON.stringify(BLANK)}`);

console.log('\n-- titleStr(): the client\'s own name wins --');
ck(titleStr({ displayName: 'Titan, Knowsley Business Park', park: 'Knowsley Business Park', unit: 'Titan' })
   === 'Titan, Knowsley Business Park', 'a stated displayName is the title, verbatim');
ck(titleStr({ displayName: BLANK, park: 'Kestrel Reach', unit: 'Unit 3' }) === 'Kestrel Reach Unit 3',
   'the BLANK sentinel is not a name, so park + unit still compose');
ck(titleStr({ displayName: 'tbd', park: 'Kestrel Reach', unit: 'Unit 3' }) === 'Kestrel Reach Unit 3',
   "...and so is a pre-v45 canonical's 'tbd'");
ck(titleStr({ displayName: '   ', park: 'Kestrel Reach' }) === 'Kestrel Reach',
   'whitespace is not a name either');
ck(titleStr({ park: 'Kestrel Reach', unit: 'Unit 3' }) === 'Kestrel Reach Unit 3',
   'the v40 park + unit rules are untouched when no displayName is stated');
ck(titleStr({ park: 'Kestrel Reach Unit 3', unit: 'Unit 3' }) === 'Kestrel Reach Unit 3',
   '...including the whole-token de-duplication');
ck(titleStr({ displayName: 'Titan' }) === 'Titan', 'a displayName alone is enough');

console.log('\n-- partyLine(): the label follows the value --');
ck(partyLine({ landlord: 'Tritax', developer: 'Panattoni' }) === 'Landlord: Tritax',
   'a stated landlord wins - it is the party to the lease');
ck(partyLine({ landlord: BLANK, developer: 'Panattoni' }) === 'Developer: Panattoni',
   'no landlord: the DEVELOPER is named, under its own label, never under "Landlord"');
ck(partyLine({ landlord: BLANK, developer: BLANK }) === `Landlord: ${BLANK}`,
   'neither stated: the label plus the blank, because "no party named" is the fact');
ck(partyLine({}) === `Landlord: ${BLANK}`, 'an empty property does not throw or print undefined');

console.log('\n-- the card --');
const titan = byPark('Knowsley Business Park');
if (!titan) { console.error('FAIL: fixture property missing'); process.exit(1); }
const card = cardHTML(titan);
ck(/<div class="dev-line">Landlord: Indurent<\/div>/.test(card),
   'the card prints the party line, and nothing else, in .dev-line');
ck(!/J\d|M\d\d?,/.test(card.split('</div>')[card.split('</div>').findIndex(x => /dev-line/.test(x))] || ''),
   'the motorway no longer rides along in that line');
ck(card.includes('Titan, Knowsley Business Park'), 'the card title is the displayName');

console.log('\n-- the modal: the four rent rows always print, every other row omits --');
const modal = detailHTML(titan);
// the label is escaped: 'Warehouse rent (monthly)' carries regex metacharacters, and an
// unescaped lookup silently returns null - which reads as "the row is absent" and would have
// passed the omission assertions below for the wrong reason.
const rx = s => s.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
const rowIn = (h, label) => {
  const m = h.match(new RegExp('<div class="spec-k">' + rx(label)
                               + '<\\/div><div class="spec-v">([\\s\\S]*?)<\\/div>'));
  return m ? m[1] : null;
};
const rowOf = (label) => rowIn(modal, label);
ck(rowOf('Warehouse rent (monthly)') !== null, 'the monthly rent row is present');
ck(rowOf('Total annual rent') !== null, 'the total annual rent row is present');
ck(rowOf('Total monthly rent') !== null, 'the total monthly rent row is present');
ck(rowOf('Warehouse rent') !== null && /9\.50/.test(rowOf('Warehouse rent')),
   'a STATED rent prints its own figure, untouched');
ck(rowOf('Service charge') === null,
   'an unstated service charge still OMITS - the always-print rule is the four rent rows only');
ck(rowOf('Lease term') === null, '...and an unstated lease term omits');
ck(rowOf('Rent-free period') === null, '...and the rent-free period');
ck(rowOf('Incentives') === null, '...and incentives');
ck(rowOf('Land price') === null, '...and land price');
ck(rowOf('REIT') === null, '...and the REIT flag, which is not a quoted term at all');
ck(rowOf('Floor load') === null, 'an unstated TECHNICAL spec still omits its row entirely');
ck(rowOf('Sprinklers') === null, '...and so does an unstated sprinkler spec');
ck(rowOf('Clear height') === '15.25m', 'a stated technical spec is unchanged');
// Asserted on the ROW VALUES rather than on a slice of the modal: the chrome's shared prose
// (coords_rings_est, modal_approx_note) carries a long dash of its own and is not this
// change's to edit, while the defect v45 fixed was five rows whose VALUE was a bare dash.
ck(!/<div class="spec-v">\s*[—–]\s*<\/div>/.test(modal),
   'no row prints a bare long dash as its value (five rows used to: three rent, two lease)');

console.log('\n-- the modal head: guarded chips, and one chip per STATED link --');
ck(/<span>Knowsley<\/span>/.test(modal), 'the city chip is there');
ck(!new RegExp('<span>' + BLANK + '</span>').test(modal),
   'no chip is printed for a field whose value is the blank token');
ck((modal.match(/class="map-link"/g) || []).length === 5,
   'five link chips: maps, brochure, video, website, Street View');
ck(modal.includes('>Watch video ↗<') && modal.includes('>Visit website ↗<')
   && modal.includes('>Street View ↗<'), 'each media chip carries its own label');
ck(!modal.includes(`href="${BLANK}"`), 'the blank token is NEVER emitted as an href');

const thin = byPark('Thin Park');
if (thin) {
  const thinModal = detailHTML(thin);
  ck((thinModal.match(/class="map-link"/g) || []).length === 1,
     'a property that states no links gets the derived maps chip and nothing else');
  ck(!/<span>undefined<\/span>/.test(thinModal), 'and no chip renders the string "undefined"');
  const rentRows = ['Warehouse rent', 'Warehouse rent (monthly)', 'Total annual rent',
                    'Total monthly rent'].map(l => rowIn(thinModal, l));
  ck(rentRows.every(v => v === BLANK),
     `a property that quotes no rent at all prints ${BLANK} on all FOUR rent rows `
     + JSON.stringify(rentRows));
  ck(!new RegExp('<div class="spec-k">Service charge<\\/div>').test(thinModal),
     '...and still omits every other commercial row');
}

console.log('\n-- Compare: every cell neutral --');
ck(!/cmp-highlight/.test(html), 'the built file carries no cmp-highlight class at all');

console.log('STATUS:', fails.length ? 'BLOCKED' : 'ALL-PASS');
process.exit(fails.length ? 1 : 0);

File: plugins/cbre-il/skills/cbre-property-longlist/evals/chrome_v45_test.py
#!/usr/bin/env python3
"""chrome_v45_test.py - the v45 chrome and pipeline change, pinned end to end.

v45 landed thirteen broker-asked changes in one version. They are pinned together because
they share two seams, and a partial revert of either seam is what would break them:

  THE BLANK TOKEN. The pipeline writes ONE token into every unfilled chrome-read field and
  the chrome prints the SAME token. Before v45 the data said 'tbd', the chrome's own
  fallbacks said 'tbd', three commercial rows said a long dash, and landPrice said a long
  dash too - four spellings of "not stated" on one page. normalize.BLANK is now the single
  owner and the chrome's `const BLANK` must equal it, so this eval reads both and compares
  them (the f05_sentinel_parity pattern: the Python side is the master, the eval holds the
  two equal). Critically the token is a MEMBER of UNKNOWN_FORMS, so every reader that asks
  looks_unknown() still reads such a field as absent - coverage, the trace gate and the
  Gaps Report are unaffected, which is the whole reason the swap is safe.

  WHAT THE CHROME NO LONGER SHOWS. Four surfaces were removed (the hero lede, the border
  POI category, Compare's best-value highlight, and its note), and a removal leaves debris
  in five places: the template, the config-token list, the i18n table, the twelve bundled
  packs, and the data the builder injects. This eval checks all five, because a key left in
  EN fails the i18n lockstep the moment someone adds a language, and a border POI left in
  the injected block is a row of data nothing on the page can render.

The executed half is the .mjs sibling: titleStr / partyLine / cardHTML / detailHTML run in
a node sandbox against a REAL built file. Everything a sandbox cannot reach is pinned here.

Offline. Drives a real build; needs node for the executed half.
Run: python evals/chrome_v45_test.py
"""
from __future__ import annotations
import json
import re
import shutil
import subprocess
import sys
import tempfile
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent
HELPERS = ROOT / "helpers"
sys.path.insert(0, str(HELPERS))
import _common as C            # noqa: E402
import build_dashboard as BD   # noqa: E402
import extract_xlsx as XL      # noqa: E402
import i18n as I18N            # noqa: E402
import normalize as N          # noqa: E402

PX = ("data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAQAAAC1HAwCAAAAC0"
      "lEQVR42mNkYPhfDwAChwGA60e6kgAAAABJRU5ErkJggg==")
LANGS = ("cs", "de", "es", "fr", "hu", "it", "nl", "pl", "pt", "ro", "sk", "zh")

FAILS: list[str] = []


def ck(ok, msg):
    print(("  [PASS] " if ok else "  [FAIL] ") + msg)
    if not ok:
        FAILS.append(msg)


def _canon():
    """Two properties: one that states everything v45 renders, one that states none of it."""
    base = {"country": "GB", "city": "Knowsley", "status": "Existing Building",
            "photo": PX, "gallery": [PX], "lat": 53.46, "lng": -2.86,
            "areaUnit": "sq ft", "rentUnit": "£/sq ft/yr"}
    rich = dict(base, **{
        "id": 1, "park": "Knowsley Business Park", "displayName": "Titan, Knowsley Business Park",
        "developer": "Indurent", "landlord": "Indurent", "warehouseArea": 184700,
        "clearHeight": "15.25m", "warehouseRent": "£9.50 / sq ft / year",
        "warehouseRentVal": 9.5, "earlyAccess": "Q1 2027", "motorway": "M57, J4 1 mile",
        "brochureLink": "https://example.com/brochure.pdf",
        "videoLink": "https://example.com/flythrough",
        "websiteLink": "https://example.com/scheme",
        "streetViewLink": "https://example.com/pano",
    })
    thin = dict(base, **{"id": 2, "park": "Thin Park", "developer": "tbd", "city": "Widnes",
                         "warehouseArea": 90000})
    return {"meta": {"client": "Matalan", "units": {"area": "sq ft", "rent": "£/sq ft/yr"},
                     "hero": {"topbar_meta": "North West · 2026-09-14", "eyebrow": "",
                              "title_html": "", "footer_copyright": "© 2026 CBRE"}},
            "pois": [{"name": "Liverpool (Port)", "type": "port", "lat": 53.43, "lng": -3.00},
                     {"name": "Swiecko (Border)", "type": "border", "lat": 52.30, "lng": 14.60},
                     {"name": "Manchester", "type": "city", "lat": 53.48, "lng": -2.24}],
            "regions": {}, "properties": [rich, thin]}


def main() -> int:                                             # noqa: C901
    tpl = (ROOT / "assets" / "dashboard_template.html").read_text(encoding="utf-8")

    print("== the BLANK token has ONE owner, and both languages read it ==")
    ck(N.BLANK == "TBC", f"normalize.BLANK is {N.BLANK!r}")
    ck(C.BLANK == N.BLANK, "_common re-exports the same object, so C.BLANK cannot drift")
    m = re.search(r'^const BLANK = "([^"]+)";', tpl, re.M)
    ck(bool(m), "the chrome declares a single BLANK constant")
    ck(bool(m) and m.group(1) == N.BLANK,
       f"the chrome's BLANK equals the pipeline's (chrome {m.group(1) if m else None!r})")
    ck(N.BLANK.lower() in N.UNKNOWN_FORMS,
       "it is a MEMBER of UNKNOWN_FORMS, so a field carrying it still reads as ABSENT")
    ck(N.looks_unknown(N.BLANK) and N.looks_unknown("tbd"),
       "both the new token and the pre-v45 one resolve to absence (no canonical is orphaned)")

    print()
    print("== the pipeline writes it, everywhere it used to write four things ==")
    ck(C.fill_render_sentinels({"status": None})["status"] == N.BLANK,
       "a required text field fills BLANK")
    ck(C.fill_render_sentinels({"clearHeight": "to be confirmed"})["clearHeight"] == N.BLANK,
       "a reader's own wording is normalised to it")
    ck(C.fill_render_sentinels({"landPrice": None})["landPrice"] == N.BLANK,
       "landPrice stops carrying a long dash of its own")
    ck(C.fill_render_sentinels({"country": None})["country"] == "??",
       "country KEEPS its code sentinel - it holds an ISO code, not prose")
    ck(C.fill_render_sentinels({"sprinklers": "None"})["sprinklers"] == "None",
       "a stated 'None' still survives (the extraction contract names it as DATA)")
    ck(N.sentinel("tbd", "landPrice") == N.BLANK and N.sentinel("tbd", "clearHeight") == N.BLANK,
       "normalize.sentinel agrees with fill_render_sentinels, landPrice included")
    ck(N.sentinel("tbd", "reit") is None, "reit is still None, not a printed token")
    ck(N.sentinel("15.25m", "clearHeight") == "15.25m", "a stated value passes through it")

    print()
    print("== the hero lede is gone, in all four places it lived ==")
    ck("{{lede}}" not in tpl, "the template carries no {{lede}} token")
    ck('class="lede"' not in tpl, "...and no .lede element")
    ck(".lede{" not in tpl and ".lede," not in tpl and ".hero .lede" not in tpl,
       "...and none of its three CSS rules")
    ck("lede" not in C.CONFIG_TOKENS, "_common.CONFIG_TOKENS no longer declares it")
    ck("hero_lede_fmt" not in I18N.EN, "the EN table no longer carries hero_lede_fmt")

    print()
    print("== the headline names the CLIENT, and never leaves a dangling separator ==")
    en = I18N.ui_for("en")
    ck("{client}" in en.get("hero_title_html", ""),
       "the EN default headline carries the {client} placeholder")
    t_named = BD._hero_copy({}, en, "Matalan")["title_html"]
    ck(t_named == "Matalan - Industrial &amp; Logistics <em>opportunities</em>",
       f"a stated client is composed in: {t_named!r}")
    ck("&amp;" in t_named and " & " not in t_named,
       "the ampersand is HTML-escaped (it lands inside an <h1>)")
    ck(t_named.count("<em>") == 1, "exactly one <em> pair - the accent colour hangs off it")
    t_blank = BD._hero_copy({}, en, "")["title_html"]
    ck(t_blank == "Industrial &amp; Logistics <em>opportunities</em>",
       f"a blank client drops the placeholder AND the separator: {t_blank!r}")
    ck("{client}" not in t_blank, "no raw placeholder ever reaches a reader")
    ck(BD._hero_copy({"title_html": "My own headline"}, en, "Matalan")["title_html"]
       == "My own headline", "an AUTHORED headline is never touched")
    ck(BD._hero_copy({}, en, "  ")["title_html"] == t_blank,
       "whitespace is not a client name")
    ck(I18N.EN.get("hero_eyebrow") == "Property Longlist",
       f"the eyebrow says Longlist, not Shortlist ({I18N.EN.get('hero_eyebrow')!r})")
    for lg in LANGS:
        pack = json.loads((ROOT / "assets" / "i18n" / f"{lg}.json").read_text(encoding="utf-8"))
        if "{client}" not in (pack.get("hero_title_html") or ""):
            ck(False, f"{lg}.json's headline lost the {{client}} placeholder")
            break
    else:
        ck(True, "all twelve packs carry the placeholder in their OWN wording")

    print()
    print("== the browser tab derives from the headline ==")
    dt = BD._doc_title({"title_html": "Matalan - Industrial &amp; Logistics <em>opportunities</em>",
                        "eyebrow": "Property Longlist"}, {"client": "Matalan"})
    ck(dt.startswith("Matalan - Industrial"), f"the headline leads, not the eyebrow: {dt!r}")
    ck("<em>" not in dt, "the tag is stripped")
    ck(dt.endswith(" · CBRE"), "the brand suffix uses the chrome's own middle dot")
    ck("—" not in dt and "–" not in dt, "and carries no long dash")
    ck(BD._doc_title({"doc_title": "Mine"}, {}) == "Mine", "an authored doc_title still wins")
    ck(BD._doc_title({}, {}) == "CBRE Property Longlist", "the last-resort default is stated")

    print()
    print("== the POI filters start OFF ==")
    ck('poiFilter: new Set()' in tpl,
       "state.poiFilter is empty, so the map opens on the properties alone")

    print()
    print("== the border category is gone from the chrome AND from the data ==")
    ck('"poi_border"' not in tpl and "poi_border" not in tpl,
       "no surface asks for the border label")
    for needle, where in (('["city","air","rail","port"]', "groupedDistances / the dist table"),
                          ('{ air:600, port:800, rail:300, city:300 }', "the live-POI caps")):
        ck(needle in tpl, f"{where} lists four categories")
    ck("cmp_nearest_border" not in tpl, "Compare has no nearest-border row")
    ck("poi_border" not in I18N.EN and "cmp_nearest_border" not in I18N.EN,
       "the two EN keys are gone")
    ck("border" not in BD.DISPLAY_POI_TYPES, "the builder does not consider it displayable")
    kept = BD._display_pois([{"type": "port"}, {"type": "border"}, {"type": "BORDER"}, {}])
    ck(len(kept) == 2 and all(str(q.get("type", "")).lower() != "border" for q in kept),
       f"_display_pois drops it case-insensitively (kept {[q.get('type') for q in kept]})")
    ck({} in kept, "a POI with NO stated type is KEPT - a filter must not become a data loss")

    print()
    print("== Compare loses the best-value highlight, and its note ==")
    ck("minSizeIdx" not in tpl and "minRentIdx" not in tpl,
       "neither index is computed any more")
    ck("cmp-highlight" not in tpl, "no cell can carry the class, and the CSS is gone too")
    ck("cmp_highlight_note" not in tpl and "cmp_highlight_note" not in I18N.EN,
       "the note and its EN key are gone")

    print()
    print("== the four new canonical fields are declared, typed and bound ==")
    schema = json.loads((ROOT / "templates" / "canonical.schema.json").read_text(encoding="utf-8"))
    sprops = schema["$defs"]["property"]["properties"]
    C._CANON_PROPERTY_FIELDS = None          # force a fresh scan (module-level cache)
    C._STRING_FIELDS_STRUCT = None
    for f in ("displayName", "videoLink", "websiteLink", "streetViewLink"):
        ck(f in sprops and sprops[f].get("type") == "string", f"the schema declares {f} as a string")
        ck(bool(sprops[f].get("description")), f"...with a description that tells a reader its rule")
        ck(f in C.canonical_property_fields(), f"{f} is a live canonical field")
        ck(f in XL.COLUMN_MAP, f"extract_xlsx binds a {f} column")
    ck("displayName" in C.STRING_FIELDS,
       "displayName is on the chrome string-field list, so an unstated one fills BLANK")
    ck(C.fill_render_sentinels({"displayName": None})["displayName"] == N.BLANK,
       "...which is what makes titleStr fall back to park + unit")
    ck(C.fill_render_sentinels({"displayName": 3})["displayName"] == "3",
       "a tracker's bare numeric name is coerced, not a schema failure")
    ck(XL._header_field("Display Name") == "displayName", "'Display Name' binds")
    ck(XL._header_field("Video") == "videoLink", "'Video' binds")
    ck(XL._header_field("Street View") == "streetViewLink", "'Street View' binds")
    ck(XL._header_field("Microsite") == "websiteLink", "'Microsite' binds")
    ck(XL._header_field("Link") != "websiteLink",
       "a bare 'Link' column stays UNBOUND rather than being guessed into one of them")
    for f in ("videoLink", "websiteLink", "streetViewLink"):
        ck(f in C.NO_TRANSLATE_FIELDS if hasattr(C, "NO_TRANSLATE_FIELDS") else True,
           f"{f} is exempt from translation (a translated URL is a broken URL)")

    print()
    print("== the availability field reads as a DATE on every surface ==")
    for key in ("row_early_access", "row_early_access_date", "cmp_early_access"):
        ck(I18N.EN.get(key) == "Available date", f"EN {key} says Available date")
    ck(I18N.EN.get("modal_early_access_prefix") == "Available",
       "the modal chip prefix is the adjective, so it reads 'Available Q1 2027'")

    print()
    print("== the i18n table and the twelve packs are in lockstep ==")
    for key in ("modal_open_video", "modal_open_website", "modal_open_streetview"):
        ck(key in I18N.EN, f"EN carries {key}")
    gone = ("hero_lede_fmt", "poi_border", "cmp_nearest_border", "cmp_highlight_note")
    bad = []
    for lg in LANGS:
        pack = json.loads((ROOT / "assets" / "i18n" / f"{lg}.json").read_text(encoding="utf-8"))
        if set(pack) != set(I18N.EN):
            bad.append("%s (%d keys vs %d)" % (lg, len(pack), len(I18N.EN)))
    ck(not bad, "every pack's key set equals EN's" + (": " + ", ".join(bad) if bad else ""))
    leftover = [f"{lg}:{k}" for lg in LANGS for k in gone
                if k in json.loads((ROOT / "assets" / "i18n" / f"{lg}.json")
                                   .read_text(encoding="utf-8"))]
    ck(not leftover, "no pack still carries a removed key" + (f": {leftover}" if leftover else ""))

    print()
    print("== a BUILT dashboard carries all of it ==")
    with tempfile.TemporaryDirectory() as td:
        d = Path(td)
        cp, hp = d / "c.json", d / "b.html"
        cp.write_text(json.dumps(_canon()), encoding="utf-8")
        BD.build(cp, hp)
        built = hp.read_text(encoding="utf-8")

        ck('class="lede"' not in built, "no lede paragraph")
        ck("Matalan - Industrial &amp; Logistics <em>opportunities</em>" in built,
           "the hero headline names the client")
        ck("<title>Matalan - Industrial &amp;amp; Logistics opportunities · CBRE</title>" in built
           or "Matalan - Industrial" in built.split("</title>")[0],
           "the browser tab does too")
        ck("(Border)" not in built.split("const PROPS")[0].split("const POIS")[-1]
           if "const POIS" in built else True, "no border POI reaches the POIS block")
        pois = json.loads(re.search(r"const POIS = (\[.*?\]);", built, re.S).group(1))
        ck(len(pois) == 2 and all(q.get("type") != "border" for q in pois),
           f"two POIs ship, neither a border crossing (got {[q.get('type') for q in pois]})")
        props = json.loads(re.search(r"const PROPS = (\[.*?\]);\s*\n", built, re.S).group(1))
        thin = [p for p in props if p["park"] == "Thin Park"][0]
        ck(thin["developer"] == N.BLANK,
           f"the unfilled fields in the DATA carry the blank token (got {thin['developer']!r})")
        ck(thin["landPrice"] == N.BLANK, "...including landPrice")
        ck("tbd" not in json.dumps(props),
           "and no property field still says 'tbd' anywhere in the shipped data")

        print()
        print("== the executed half (node) ==")
        node = shutil.which("node")
        if not node:
            ck(False, "node is required to execute the chrome (install node or add it to PATH)")
        else:
            mjs = Path(__file__).with_suffix(".mjs")
            r = subprocess.run([node, str(mjs), str(hp)], capture_output=True, text=True,
                               timeout=180)
            print(r.stdout.rstrip())
            if r.stderr.strip():
                print(r.stderr.rstrip())
            ck(r.returncode == 0, "the EXECUTED chrome assertions pass (see above)")

    print()
    if FAILS:
        print(f"STATUS: BLOCKED ({len(FAILS)} failure(s))")
        return 1
    print("STATUS: ALL-PASS")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
